setuptools/discovery.py

```diff
@@ -148,6 +151,13 @@ def _find_iter(cls, where: _Path, exclude: _Filter, include: _Filter) -> StrIter
 
                 # Should this package be included?
                 if include(package) and not exclude(package):
+                    top_package = package.split(".")[0]
+                    if top_package in FlatLayoutPackageFinder._EXCLUDE:
+                        warnings.warn(
+                            f"{top_package!r} is a reserved package name. "
+                            "Please add it to find_package(exclude=...)",
+                            setuptools.SetuptoolsDeprecationWarning,
+                        )
```

Comment:
Is it really the case these patterns should be deprecated?

What if the user is the owner of the package `tasks` or `changelog` (those are perfectly valid entries for PyPI right?) and is explicitly using `[options.packages.find] include = tasks*`?

I am also not sure if this is the best place to implement the check..
This code path is traversed by the auto-discovery algorithm that will already error if the distribution contains multiple top-level packages... It is also traversed when the `src-layout` is being used (and my opinion is that if the `src-layout` is used, there should be no warning...).

I think that an alternative would be modifying `setuptools.find_packages`, `setuptools.find_namespace_packages`, `setuptools.config.setupcfg.ConfigOptionsHandler._parse_packages` and `setuptools.pyprojecttoml._ConfigExpander._expand_packages` to check for one of these patterns ***only if*** there is more than one top-level package.